snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedConceptRequestCache.java

```diff
@@ -0,0 +1,124 @@
+import com.b2international.snowowl.core.domain.IComponent;
+import com.b2international.snowowl.snomed.core.domain.SnomedConcept;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSortedSet;
+
+/**
+ * @since 8.1.0
+ */
+public class SnomedConceptRequestCache {
+
+	private final Map<FetchConfig, Map<String, SnomedConcept>> cache = new HashMap<>(); 
+	private final Deque<FetchConfig> requestedFetches = new ArrayDeque<>();
+	
+	public void request(Iterable<String> ids, Options expand, List<ExtendedLocale> locales, Consumer<Map<String, SnomedConcept>> onConceptsReady) {
+		requestedFetches.add(new FetchConfig(ids, expand, locales, onConceptsReady));
+	}
+	
+	public Map<String, SnomedConcept> get(BranchContext context, Iterable<String> ids, Options expand, List<ExtendedLocale> locales) {
+		FetchConfig config = new FetchConfig(ids, expand, locales, concepts -> {});
+		if (cache.containsKey(config)) {
+			return cache.get(config);
+		} else {
+			requestedFetches.add(config);
+			compute(context);
+			return cache.get(config);
+		}
+	}
+	
+	public void compute(BranchContext context) {
+		// run until all requested fetches are resolved
+		while (!requestedFetches.isEmpty()) {
+			// using the currently first requestedFetch config
+			// check if there are similar fetchConfigs requested earlier, and if yes, try to fetch them now, so they can be referenced later from the cache, and immediately get populated via the callback
+			FetchConfig config = requestedFetches.getFirst();
+			final Set<FetchConfig> configsToFetch = requestedFetches.stream()
+					.filter(cfg -> Objects.equals(config.expand, cfg.expand) && Objects.equals(config.locales, cfg.locales))
+					.collect(Collectors.toSet());
+			// remove them from the requested fetches
+			requestedFetches.removeAll(configsToFetch);
+
+			// search for configs using the same expand and locales, merge the IDs and fetch all of them together
+			final Set<String> ids = configsToFetch.stream().flatMap(cfg -> cfg.ids.stream()).collect(Collectors.toSet());
+			
+			context.log().info("Fetching concepts: ids={}, expand={}, locales={}", ids, config.expand, config.locales);
+			Map<String, SnomedConcept> fetchedConcepts = SnomedRequests.prepareSearchConcept()
+					.setLimit(ids.size())
+					.filterByIds(ids)
+					.setExpand(config.expand)
+					.setLocales(config.locales)
+					.build()
+					.execute(context)
+					.stream()
+					.collect(Collectors.toMap(IComponent::getId, c -> c));
+			
+			// populate the cache for each fetch config and call the callback
+			configsToFetch.forEach(cfg -> {
+				ImmutableMap.Builder<String, SnomedConcept> configConcepts = ImmutableMap.builder();
+				cfg.ids.forEach(idToCache -> {
+					configConcepts.put(idToCache, fetchedConcepts.get(idToCache));
+				});
+				ImmutableMap<String, SnomedConcept> concepts = configConcepts.build();
+				cache.put(cfg, concepts);
+				cfg.onConceptsReady.accept(concepts);
```

Comment:
Callbacks could also use `fetchedConcepts` directly, which contains a superset of the concepts they have originally requested. Then you don't need to build a map for each consumer.

Reply:
Yeah, thought about that, it would work in certain cases where we only use the map by requested concepts via their ID from it, but in case you iterate over all of the entries of the returned map, this would give you inconsistent results. `config.ids` and `map.keys` should be equal.